File: packages/showrunner/src/renderer/components/updates/update-status-view.test.ts
import { describe, expect, it } from "vitest"
import type { UpdateStatus } from "showrunner-schema"
import { friendlyUpdateError, getUpdateStatusView } from "./update-status-view"

const baseStatus: UpdateStatus = {
	currentVersion: "1.0.0",
	hasUpdate: false,
	canCheckForUpdates: true,
	checking: false,
}

describe("update status view", () => {
	it("shows development build state when update checks are unavailable", () => {
		expect(
			getUpdateStatusView({
				...baseStatus,
				canCheckForUpdates: false,
				message: "Update checks run in packaged builds.",
			})
		).toMatchObject({
			statusTitle: "Development build",
			statusDetail: "Update checks run in packaged builds.",
			statusClass: "updates-page__status--muted",
		})
	})

	it("shows offline update errors without exposing low-level network noise", () => {
		expect(friendlyUpdateError("connect ETIMEDOUT github.com")).toBe(
			"Could not reach the update server. Check your internet connection and try again."
		)
	})

	it("shows no-update state after a successful check", () => {
		expect(getUpdateStatusView({ ...baseStatus, checkedAt: "2026-05-03T12:00:00.000Z" })).toMatchObject({
			statusTitle: "You're up to date",
			statusDetail: "ShowRunner v1.0.0 is the current installed version.",
			statusClass: "updates-page__status--current",
		})
	})

	it("shows update-available state", () => {
		expect(
			getUpdateStatusView({
				...baseStatus,
				hasUpdate: true,
				update: { version: "1.0.1", name: "Release", date: "", notes: "" },
			})
		).toMatchObject({
			statusTitle: "Update available",
			statusDetail: "v1.0.0 -> v1.0.1",
			latestVersionLabel: "v1.0.1",
		})
	})

	it("shows downloaded state before generic available state", () => {
		expect(
			getUpdateStatusView({
				...baseStatus,
				hasUpdate: true,
				downloaded: true,
				update: { version: "1.0.1", name: "Release", date: "", notes: "" },
			})
		).toMatchObject({
			statusTitle: "Update downloaded",
			statusDetail: "The update is downloaded and ready to install on restart.",
		})
	})
})

File: packages/showrunner-flutter/test/features/support/update_status_view_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:showrunner_flutter/features/support/update_status_view.dart';
import 'package:showrunner_flutter/schema/update.dart';

void main() {
  test('shows unknown latest version before the first check', () {
    final view = updateStatusView(
      const UpdateInfo(
        currentVersion: '2.0.0',
        latestVersion: '2.0.0',
        hasUpdate: false,
      ),
    );

    expect(view.title, 'Ready to check');
    expect(view.latestVersionLabel, 'unknown');
    expect(view.tone, UpdateStatusTone.current);
  });

  test(
    'shows the installed and latest versions when an update is available',
    () {
      final view = updateStatusView(
        const UpdateInfo(
          currentVersion: '2.0.0',
          latestVersion: '2.1.0',
          hasUpdate: true,
          status: UpdateStatus.available,
          checkedAt: '2026-09-23T09:00:00Z',
        ),
      );

      expect(view.title, 'Update available');
      expect(view.detail, 'v2.0.0 → v2.1.0');
      expect(view.latestVersionLabel, 'v2.1.0');
      expect(view.tone, UpdateStatusTone.available);
    },
  );

  test('shows a concise offline error and keeps latest version unknown', () {
    final view = updateStatusView(
      const UpdateInfo(
        currentVersion: '2.0.0',
        latestVersion: '2.0.0',
        hasUpdate: false,
        status: UpdateStatus.error,
        errorMessage: 'Update check is offline.',
        checkedAt: '2026-09-23T09:00:00Z',
      ),
    );

    expect(view.title, 'Update check failed');
    expect(view.detail, contains('internet connection'));
    expect(view.latestVersionLabel, 'unknown');
    expect(view.tone, UpdateStatusTone.error);
  });
}
